src/primitives/decode/mod.rs

```diff
@@ -0,0 +1,384 @@
+    /// and checksum).
+    pub fn fe32_iter(&self) -> Fe32Iter {
+        Fe32Iter { iter: AsciiToFe32Iter { iter: self.data.iter().copied() } }
+    }
+
+    /// Parses an bech32 encode string and constructs a [`Parsed`] object.
+    ///
+    /// Checks for valid ASCII values, does not validate the checksum.
+    ///
+    /// The object returned by this function must must have `validate_checksum` called on it to
+    /// validate and checksum and maintain data invariant (remove checksum from parsed data).
+    fn parse_unvalidated(s: &'s str) -> Result<Self, Error> {
+        let sep_pos = check_characters(s)?;
+        let (hrp, data) = s.split_at(sep_pos);
+
+        let ret = Parsed {
+            hrp: Hrp::parse(hrp)?,
+            data: data[1..].as_bytes(), // Skip the separator.
+        };
+
+        Ok(ret)
+    }
+
+    /// Validates that a [`Parsed`] returned by `parse_unvalidated` has a valid checksum.
+    ///
+    /// # Returns
+    ///
+    /// Returns `self` with the checksum removed from the inner data slice.
+    fn validate_checksum<Ck: Checksum>(mut self) -> Result<Self, Error> {
+        if Ck::CHECKSUM_LENGTH == 0 {
+            return Ok(self); // Called with NoChecksum.
+        }
+
+        if self.data.len() < Ck::CHECKSUM_LENGTH {
+            return Err(Error::InvalidChecksumLength);
+        }
+
+        let mut checksum_eng = checksum::Engine::<Ck>::new();
+        checksum_eng.input_hrp(&self.hrp());
+
+        // Unwrap ok since we checked all characters in our constructor.
+        for fe in self.data.iter().map(|&b| Fe32::from_char_unchecked(b)) {
+            checksum_eng.input_fe(fe);
+        }
+
+        if checksum_eng.residue() != &Ck::TARGET_RESIDUE {
+            return Err(Error::InvalidChecksum);
+        }
+
+        let data_len = self.data.len() - Ck::CHECKSUM_LENGTH;
+        self.data = &self.data[..data_len];
+
+        Ok(self)
+    }
+}
+
+/// A iterator over a parsed HRP string data as bytes.
+pub struct ByteIter<'s> {
+    iter: FesToBytes<AsciiToFe32Iter<iter::Copied<slice::Iter<'s, u8>>>>,
+}
+
+impl<'s> Iterator for ByteIter<'s> {
+    type Item = u8;
+    fn next(&mut self) -> Option<u8> { self.iter.next() }
+    fn size_hint(&self) -> (usize, Option<usize>) { self.iter.size_hint() }
+}
+
+/// A iterator over a parsed HRP string data as field elements.
+pub struct Fe32Iter<'s> {
+    iter: AsciiToFe32Iter<iter::Copied<slice::Iter<'s, u8>>>,
+}
+
+impl<'s> Iterator for Fe32Iter<'s> {
+    type Item = Fe32;
+    fn next(&mut self) -> Option<Fe32> { self.iter.next() }
+    fn size_hint(&self) -> (usize, Option<usize>) { self.iter.size_hint() }
+}
+
+/// Helper iterator adaptor that maps an iterator of valid bech32 character ASCII bytes to an
+/// iterator of field elements.
+///
+/// This iterator is a performance optimization. Equivalent, but significantly faster than, using
+/// `hrp_string.data_chk.iter().copied().map(|c| Fe32::from_char_unchecked(c).to_u8())`.
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

This comment is out of date and can probably just be dropped. There is no `data_chk` field on `Parsed`.

Reply:
Removed.